pkg/source/adapter/adapter.go

```diff
@@ -208,11 +227,100 @@ func (a *Adapter) SetRateLimits(r rate.Limit, b int) {
 	a.rateLimiter = rate.NewLimiter(r, b)
 }
 
-func (a *Adapter) HandleServiceMessage(ctx context.Context, message ctrl.ServiceMessage) {
-	// In this first PR, there is only the RA sending messages to control plane,
-	// there is no message the control plane should send to the RA
-	a.logger.Info("Received unexpected control message")
+func (a *Adapter) HandleSetContract(ctx context.Context, message ctrl.ServiceMessage) {
+	newContract := kafkasourcecontrol.KafkaSourceContract{}
+	if err := newContract.UnmarshalBinary(message.Payload()); err != nil {
+		a.logger.Errorf("Catastrophic failure: received an unparsable contract. This sounds like a programming bug: %v", err)
+		message.Ack()
+		return
+	}
+
+	// We have the contract, we can ack back now
 	message.Ack()
+
+	// Let's lock and proceed with the changes
+	a.reconfigurationLock.Lock()
+	defer a.reconfigurationLock.Unlock()
+
+	if reflect.DeepEqual(&newContract, a.actualContract) {
+		// Nothing to do here
+		return
+	}
+
+	a.stopConsumerGroup()
+
+	a.actualContract = &newContract
+
+	consumerContext, stopConsumer := context.WithCancel(ctx)
+	a.stopConsumer = stopConsumer
+	var startupError error
+	a.consumerStoppedSignal, startupError = a.startConsumerGroup(consumerContext, consumer.WithSaramaConsumerLifecycleListener(a))
+
+	updateResult := kafkasourcecontrol.UpdateResult{Generation: newContract.Generation}
+	if startupError != nil {
+		updateResult.Error = fmt.Sprintf("Error while starting the consumer group: %v", startupError)
+	}
+	if err := a.controlServer.SendAndWaitForAck(kafkasourcecontrol.NotifyContractUpdated, updateResult); err != nil {
+		a.logger.Errorf("Catastrophic failure: cannot contact the controller to notify the contract update: %v", err)
+	}
```

Comment:
I'm looking at how to abstract this common pattern to send commands and then send an async reply with success or failure